### ISSUE TITLE
Specs with Mock started failing on v5.16.0

### ISSUE THREAD

**Issue by baelter** [minitest/minitest#912]
Don't have a lot of information yet. 

But some specs using `MiniTest::Mock` starting failing after upgrading minitest.

Will update when I figure out what's going on.

**Comment by baelter** [comment]
Likely due to changes in argument handling in `mock.rb`

Seeing: `ArgumentError: mocked method :my_method? expects 2 arguments, got [:arg1]`

But I am calling it with one argument and one keyword argument.

**Comment by N0xFF** [comment]
How we can allow any `kwargs`?

In 5.15 it was possible with `Hash` class:

```ruby
mock.expect(:my_method, 42, [Hash])
```

**Comment by baelter** [comment]
With this change kwargs should not be in the args array anymore:
https://github.com/minitest/minitest/compare/v5.15.0...v5.16.0#diff-2a72d0121a0eb22e9f32a1ce654e5132391273007bf643b5ae68fbbc2800dea4L81
But maybe that should be fixed since it's a breaking change?

**Comment by sedubois** [comment]
We also had a broken test when updating from 5.15.0 to 5.16.0. We use MiniTest::Mock through Rails ActiveSupport [`assert_called_with`](https://github.com/rails/rails/blob/7-0-stable/activesupport/lib/active_support/testing/method_call_assertions.rb#L20), which does not seem to offer a way to pass kwargs separately.

**Comment by eileencodes** [comment]
@sedubois `assert_called_with` comes from Rails not minitest. However is a private API. I'm working on fixing it for our tests but it shouldn't be consumed by applications as I do not intend to maintain the existing method signature or behavior.

**Comment by zenspider** [comment]
I didn't realize that by adding real kwsupport I'd break people's kludged keywords... how were you doing it? I assume a hash on the end of args? Can someone provide some failing examples? This issue is not workable as-is.

**Comment by zenspider** [comment]
@eileencodes lemme know if there's anything I can do to help you on your end.

**Comment by thomthom** [comment]
I ran into this one as well. In my case it all related to testing `Observer` logic from the std lib;

```
C:/Ruby27-x64/lib/ruby/2.7.0/observer.rb:197: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
C:/projects/extension-sources/vendor/bundle/ruby/2.7.0/gems/minitest-5.16.0/lib/minitest/mock.rb:31: warning: The called method is defined here
```

I'll try to distil a minimal example.

**Comment by thomthom** [comment]
Got a minimal repo. In my case the issue is that I have some objects that implement `to_hash`. When these objects are passed as arguments to `Observerable`s `notify_observers` call there's an error thrown:

```rb
require 'minitest/autorun'

require 'observer'

class Example

  include Observable

  def trigger
    changed
    notify_observers(:triggered, self)
  end

  # Remove this method and the test will pass.
  # In Ruby 2.7 this cause the mock-observer interaction to fail with MiniTest 5.16.0.
  def to_hash
    { foo: 'bar' }
  end

end

class ExampleTest < Minitest::Test

  def test_example_trigger
    observable = Example.new
    mock = Minitest::Mock.new
    mock.expect(:hash, mock.object_id)
    mock.expect(:event, nil, [:triggered, observable])
    observable.add_observer(mock, :event)
    observable.trigger
  end

end
```

Ruby 2.7 and MiniTest 5.16.0:
```
  1) Error:
ExampleTest#test_example_trigger:
ArgumentError: mocked method :event expects 2 arguments, got [:triggered]
    C:/Ruby27-x64/lib/ruby/2.7.0/observer.rb:197:in `block in notify_observers'
    C:/Ruby27-x64/lib/ruby/2.7.0/observer.rb:196:in `each'
    C:/Ruby27-x64/lib/ruby/2.7.0/observer.rb:196:in `notify_observers'
    C:/Users/Thomas/source/example/tests/standalone/model/example_test.rb:11:in `trigger'
    C:/Users/Thomas/source/example/tests/standalone/model/example_test.rb:30:in `test_example_trigger'
```

In Ruby 2.5 and Ruby 3.1 that example works just fine.

If I remove the `to_hash` method (not an option) then Ruby 2.7 also passes.

Previous versions of MiniTest also worked fine with the above example.

Looks like something in Ruby 2.7 and MiniTest 5.16.0 causes the `to_hash`'able object to be resolves incorrectly when passed as an argument. There was some changes in Ruby to how hashes and keyword params were resolves, weren't there?

**Comment by zrzka** [comment]
Just experienced it as well. Ruby 3.1.2, Rails 6.1.6 and MiniTest 5.16.0. 5.15.0 works fine.

```
DockerhubPublisherTest#test_publish_multi_arch:
--
  | ArgumentError: mocked method :tag expects 1 arguments, got []
  | lib/licensor/dockerhub_publisher.rb:192:in `block in tag'
  | lib/licensor/dockerhub_publisher.rb:190:in `map'
  | lib/licensor/dockerhub_publisher.rb:190:in `tag'
  | lib/licensor/dockerhub_publisher.rb:78:in `publish_multi_arch_image'
  | lib/licensor/dockerhub_publisher.rb:56:in `publish'
```

It's caused by:

```
mock.expect :tag, true, [{ repo: 'foo/bar', tag: "1.0.0-#{SomeModule::DockerhubPublisher::X86_SUFFIX}" }]
```

Called in the code as:

```
image.tag(repo: repo, tag: tag)
```

**Comment by baelter** [comment]
Think you need something like
If mocked method accepts kwags
and no kwargs passed to expect
and last arg of expect args is hash
pop last arg as kwarg and use to call mock

**Comment by dazuma** [comment]
We also just ran into this for some of Google's libraries. In 5.15, when mocking a method with keyword arguments, we'd need to pass the keyword arguments as a hash at the end of the positional arguments array passed to `Mock#expect`. Now in 5.16, that usage no longer works, and we need to pass the keyword arguments separately. So it's effectively a breaking change. The good news is, the new usage does work across Ruby 2 and 3.

**Comment by zenspider** [comment]
@baelter unfortunately, you have no idea what the real method takes at the point where `expect` is called.

And I'm in the sad position of looking at the last arg and not being sure if the user really wanted to pass a hash or if they wanted kwargs... Maybe a warning for a while?

**Comment by thomthom** [comment]
What would the warning be?

Would behaviour remain as in 5.16? (How would you then mock a method call where the only or last item is a Hash?)

**Comment by zenspider** [comment]
I now have a fix for the old-kwargs scenario as shown by @zrzka ... I don't yet have a fix for 2.6/2.7 for the observer issue from @thomthom yet (but I haven't investigated, so I'm not sure it is out of my hands or not).

Current fix: Set `MT_KWARGS_HACK=1` to get the old scenario to work.  If you set that var AND have real kwords AND have a hash at the end, it'll warn (once)... so it should only be an issue as you're transitioning over to the new setup.

**Comment by zenspider** [comment]
OK. I've investigated the observer scenario. Because 2.7 uses `send` without a `**kwargs` it is going to be shoving kwargs into a plain `*args` and then complaining about it. The good news is in both `MT_KWARGS_HACK=1` and without, the observer code warns but passes.... So I'm gonna chalk that up as unfortunate, but working fine. The workarounds are to either upgrade past 2.7 or not declare your class as "is-a Hash" by having a `to_hash` method (consider having `to_h` instead). If it works, but warns, it's still working.

**Comment by zenspider** [comment]
I misstated the observer scenario above. With the code I just pushed, here's the outcomes:

```ruby
    # without ENV:
    mock.expect(:event, nil, [:triggered, observable])   # original: fails
    mock.expect(:event, nil, [:triggered, **observable]) # use to_hash: fails
    mock.expect(:event, nil, [:triggered], observable)   # kwargs: pass
    mock.expect(:event, nil, [:triggered], **observable) # kwargs/to_hash: pass

    # with ENV MT_KWARGS_HACK=1
    mock.expect(:event, nil, [:triggered, observable])   # original: fails
    mock.expect(:event, nil, [:triggered, **observable]) # use to_hash: pass
    mock.expect(:event, nil, [:triggered], observable)   # kwargs: pass
    mock.expect(:event, nil, [:triggered], **observable) # kwargs/to_hash: pass
```

I doubt I'm going to come up with a working version that deals with: ruby 2.7's observer + Example w/ to_hash + minitest 5.16 with or without the ENV adjustment above. I think it's just gonna have to be modified in 1 of the 3 ways above. 

Given that the problem goes away with either: not treating your class as a is-a-Hash or using ruby 3.0+ or using your expectations in a more complaint way... 

... well let's just say I'm open to a compelling argument for more compatibility, I just doubt it'll be that compelling at this point.

**Comment by zenspider** [comment]
I have a rule about not releasing after midnight ESPECIALLY if liquor has been involved... but this fix will come out soon enough.

**Comment by thomthom** [comment]
Thanks for investigating that. I might not really need to use `to_hash` and could possibly use `to_h` instead. (I need to double check my JSON serialization logic.)

**Comment by zenspider** [comment]
to_hash is like to_ary and friends (vs to_h or to_a)... Only should be used if they're really implicitly "is-a"... So def worth considering

**Comment by zenspider** [comment]
@casperisfine and this doesn't work using the env? Can you scratch me up a quick test? You can steal a skeleton from 2-3 commits ago.

**Comment by casperisfine** [comment]
> this doesn't work using the env?

I haven't tried, I just rewrote these mocks using `mocha`.

> Can you scratch me up a quick test? 

👀

**Comment by casperisfine** [comment]
Here you go: https://github.com/Shopify/minitest/commit/a527a74bd75f1450ad17675a75b7af483f053125, apparently it doesn't work with the env either.

**Comment by eileencodes** [comment]
I have another example of code that breaks with teh upgrade in Ruby 2.7 (works fine in 3.0+) and doesn't work with the ENV var hack. My PR is here https://github.com/rails/rails/pull/45370, failures here https://buildkite.com/rails/rails/builds/87503#01818819-5daf-4cd7-bdfb-e48ec43c837c and you can run one of the Railties tests with `bundle exec ruby -Ilib:test test/generators/actions_test.rb -n test_rails_command_with_env_option_should_run_rails_with_the_env_environment`. Let me know if it's easier to debug outside of Rails and I'll try to make a demo app or script that repros.

**Comment by tsugimoto** [comment]
We are also stuck in this problem.
From Minitest 5.16.0, the last argument of the stubbed method should not be unpermitted `ActionController::Parameters` because Ruby 2.7 tries to convert it into kwargs with `#to_h` and raises.

```
ActionController::UnfilteredParameters: unable to convert unpermitted parameters to hash
from /usr/local/bundle/gems/actionpack-6.1.6/lib/action_controller/metal/strong_parameters.rb:316:in `to_h'
```

For example, the method `Auth::Token::create_or_retry`

```rb
class Auth::Token < ApplicationRecord
  def self.create_with_retry(user, raw_params, limit: 3)
    # actually create a token with retry, and record generic information from raw_params
  end
end
```

with this usage

```rb
class Auth::TokensController < ApplicationController
  def create
    device = Auth::Token.create_with_retry(current_user, params)
  end
end
```

couldn't be stubbed with any of these in Ruby 2.7 (some of them work in Ruby 3.0+)

```rb
Auth::Token.stub(:create_with_retry, ->(_user, _raw_params)
Auth::Token.stub(:create_with_retry, ->(_user, _raw_params, limit:)
Auth::Token.stub(:create_with_retry, ->(_user, _raw_params, **kwargs)
```

**Comment by zenspider** [comment]
@casperisfine I have your test in and passing...

**Comment by zenspider** [comment]
@tsugimoto `stub` isn't `mock` so please file a separate issue

**Comment by casperisfine** [comment]
> I have your test in and passing...

It's failing on CI: https://github.com/Shopify/minitest/runs/6999031451?check_suite_focus=true#step:5:477

**Comment by miguelperez** [comment]
> OK. This is released in 5.16.1. Please update, `export MT_KWARGS_HACK=1`, and see how that works for y'all.
> 
> _(and **please** only use this workaround in order to have working tests while you transition to proper kwargs calling conventions.)_

Hey, thanks for the update. I would like to know what are `"proper kwargs calling conventions"`.

For example, tests implemented following this https://www.reinhardt.io/ruby/testing/2021/05/25/stubbing-and-mocking-in-minitest.html started to fail.

```ruby
test "it will set the type to load" do
  mock = Minitest::Mock.new

  args_checker = lambda do |args|
    assert_equal :load, args[:type]
  end
  mock.expect :notify!, true, [args_checker]

  SomeClass.stub(:new, mock) do
    logic_that_triggers_notify
  end

  mock.verify
end

# method in question is defined like this:
def notify!(type:)
  something
end
```

The number of tests that fail is small < 5. So I can change them but would like to understand better whether the tests could be specified in a more proper way.

**Comment by zenspider** [comment]
@eileencodes yeah... i'mma need some help untangling that patch into a test I can debug

**Comment by zenspider** [comment]
@eileencodes this will work (once released):

https://github.com/rails/rails/pull/45370/files#diff-14cab604508ced2a334cad4623993cac14e1d4b55d6b4e1bc681b830924f492aL178

**Comment by zenspider** [comment]
@eileencodes BUT!I don't want to release until I get a thumbs up from you (or a thumbs down and smaller repro)

**Comment by eileencodes** [comment]
> BUT!I don't want to release until I get a thumbs up from you (or a thumbs down and smaller repro)

It doesn't look like it's fixed, but I don't know how to get the unreleased version hooked up to Rails CI to do a full test it since there's no gemspec. I did this locally by creating a gemspec manually but that's not going to work on Rails CI without forking, which I guess I can do to see the full set of failures.

Is there another way to do this more easily? I've never used hoe so I'm not sure how to make it play nicely with bundler's expectations.

**Comment by zenspider** [comment]
@eileencodes I really just need a failing test extracted enough that I don't need the whole world to come with it. I _think_ I've done it. I've got a run w/:

```
  1) Error:
ActiveSupport::Cache::RedisCacheStoreTests::InitializationTest#test_stuff:
ArgumentError: mocked method :call expects 1 arguments, got []
```

with the following:

```ruby
require "minitest/autorun"

class Redis
end

class ActiveSupport
  class Cache
    class RedisCacheStore
      attr_accessor :kws
      def initialize(**kws)
        self.kws = kws
      end

      def redis
        call(:url                => "redis://localhost:6379/0",
             :connect_timeout    => 20,
             :read_timeout       => 1,
             :write_timeout      => 1,
             :reconnect_attempts => 0,
             :driver             => "hiredis")
      end

      def call **kws
        Redis.new
      end
    end
  end
end

module ActiveSupport::Cache::RedisCacheStoreTests
  REDIS_URL  = "redis://localhost:6379/0"
  REDIS_URLS = %w[ redis://localhost:6379/0 redis://localhost:6379/1 ]
  REDIS_UP   = true
  DRIVER     = "hiredis"

  class InitializationTest < Minitest::Test
    def build(**kwargs)
      ActiveSupport::Cache::RedisCacheStore.new(driver: DRIVER, **kwargs.merge(pool: false)).tap(&:redis)
    end

    def test_stuff
      default_args = {
        connect_timeout: 20,
        read_timeout: 1,
        write_timeout: 1,
        reconnect_attempts: 0,
        driver: DRIVER
      }

      mock = Minitest::Mock.new
      mock.expect(:call, Redis.new, [{ url: REDIS_URLS.first }.merge(default_args)])
      mock.expect(:call, Redis.new, [{ url: REDIS_URLS.last }.merge(default_args)])

      Redis.stub(:new, mock) do
        @cache = build url: REDIS_URLS
        assert_kind_of ::Redis, @cache.redis
      end

      assert_mock(mock)
    end
  end
end
```

BUT! this fails across the board, regardless of version it seems. Found it here:

https://buildkite.com/rails/rails/builds/87503#01818819-5d9b-439d-ba67-f905c797161c/1169-1202

can you verify that the test extraction is close enough for the repro you're seeing? It just needs to be able to run against a local copy (or install) of minitest and reproduce the error.

**Comment by eileencodes** [comment]
> I really just need a failing test extracted enough that I don't need the whole world to come with it. I think I've done it. I've got a run w/:

I understand but I wanted to know all of the failures so I can give you an example of the different sets of failures we have rather than having to go back and forth if one set is fixed and another is not.

**Comment by zenspider** [comment]
@eileencodes PS activesupport test above runs fine w/ no extra `-I` or `bundle exec`... activemodel does not. I found that AS uses require_relative and the helper requires bundler/setup whereas AM does not. I'm not really wanting to clean all that up in this PR but I can put a small start in for the test file I was trying.

**Comment by zenspider** [comment]
I haven't heard anything back on the rails side... I'm going to release what I've got.

**Comment by zenspider** [comment]
This is released in 5.16.2. Thanks for reporting! @baelter, please close if this addresses your issues.

Others: please open a NEW issue if you're edge cases aren't addressed.

**Comment by baelter** [comment]
Works when changing to the new method signature, so that is good. We can lift the pin and update our specs. `MT_KWARGS_HACK=1` also works fine in our case. Nice job!
